File: dsp_modules/modules/flanger/test/flanger_test.c
#include <stdio.h>
#include "../../../test-tools/c/fopnchk/fopnchk.h"
#include <modules.h>

int main() {      
    printf("\n");
    printf("-------------\n");
    printf("-- TESTING --\n");
    printf("-------------\n");  

    float rate = 0.9;
    float delay = 0.015;
    int sampling_rate = 16000;

     // READ FILE
    FILE *f_in;
    f_in = fopen("../../../test-tools/signals/cleanguitar_sample.f32","rb");
    if (!f_in) {
        printf("Unable to open input file!");
        return -1;
    }  

    FILE *f_out = fopnchk("testvectors/c/output.f32","wb");
    
    FILE *f_in_out = fopnchk("testvectors/c/in.f32","wb");
    FILE *f_delaybuf = fopnchk("testvectors/c/delaybuffer.f32","wb");
    FILE *f_wr_ptr = fopnchk("testvectors/c/wr_ptr.s32","wb");
    FILE *f_sin_term = fopnchk("testvectors/c/sin_term.f32","wb");
    FILE *f_pi_cnt = fopnchk("testvectors/c/pi_cnt.s32","wb");

    FILE *f_sin_inc = fopnchk("testvectors/c/sin_inc.f32","wb");
    FILE *f_pi_reset = fopnchk("testvectors/c/pi_reset.s32","wb");
    FILE *f_time_shift_idx = fopnchk("testvectors/c/time_shift_idx.s32","wb");
    FILE *f_n = fopnchk("testvectors/c/n.s32","wb");
    

    t_flanger_data flanger_data;
    int init_code = flanger_init(&flanger_data, BLOCKSIZE,sampling_rate, rate, delay);
    if (init_code!=0) printf("Init error code: %i\n", init_code);

    float inbuf[BLOCKSIZE];
    while (fread(inbuf, sizeof(float), BLOCKSIZE, f_in) == BLOCKSIZE)
    {
        flanger_apply(&flanger_data, inbuf);

        fwrite(&flanger_data.output, sizeof(float), BLOCKSIZE, f_out);

        fwrite(inbuf,                       sizeof(float),  BLOCKSIZE,              f_in_out);
        fwrite(&flanger_data.delay_buffer,  sizeof(float),  FLANGER_DELAYLINE_SIZE, f_delaybuf);
        fwrite(&flanger_data.wr_ptr,        sizeof(int),    1,                      f_wr_ptr);
        fwrite(&flanger_data.sin_term,      sizeof(float),  1,                      f_sin_term);
        fwrite(&flanger_data.pi_cnt,        sizeof(int),    1,                      f_pi_cnt);

        fwrite(&flanger_data.pi_reset,        sizeof(int),    1,                      f_pi_reset);
        fwrite(&flanger_data.sin_inc,        sizeof(int),    1,                      f_sin_inc);

fwrite(&flanger_data.time_shift_idx,        sizeof(int),    1,                      f_time_shift_idx);
fwrite(&flanger_data.n,        sizeof(int),    1,                      f_n);

    }

    fopnchk_close_all();    
    return 0;
}

File: dsp_modules/modules/flanger/test/flanger_test.m
clear all
BLOCKSIZE = 48;
FS = 16000;

addpath('../matlab');

% run and write testvector to file
if ~exist('testvectors/matlab','dir')
  mkdir('testvectors/matlab');
end
fo0 = fopen('testvectors/matlab/output.f32','wb');

fo_pi_cnt = fopen('testvectors/matlab/pi_cnt.s32','wb');
fo_sin_term = fopen('testvectors/matlab/sin_term.f32','wb');
fo_wr_ptr = fopen('testvectors/matlab/wr_ptr.s32','wb');
fo_delay_buf = fopen('testvectors/matlab/delaybuffer.f32','wb');
fo_in_out = fopen('testvectors/matlab/in.f32','wb');
fo_pi_reset = fopen('testvectors/matlab/pi_reset.s32','wb');
fo_sin_inc = fopen('testvectors/matlab/sin_inc.f32','wb');
fo_time_shift_idx = fopen('testvectors/matlab/time_shift_idx.s32','wb');
fo_n = fopen('testvectors/matlab/n.s32','wb');

[in, fs0] = audioread('../../../test-tools/signals/cleanguitar_sample.wav');
% file_name = 'acoustic.wav';
% file_name = 'cleanguitar_morebass.wav';
% file_name = '.wav';
% [in,fs] = audioread(file_name);

rate = 0.9;
delay = 0.015;
o = flanger(BLOCKSIZE, FS, rate, delay);

number_of_blocks = floor(length(in) / BLOCKSIZE);
output = [];

pi_cnt = [];
sin_term = [];
wr_ptr = [];
delay_buf = [];
in_out = [];
pi_reset = [];
sin_inc = [];
time_shift_idx = [];
n = [];

for i = 1:number_of_blocks   
    
    o.flanger_apply(in(i*BLOCKSIZE - BLOCKSIZE + 1:i*BLOCKSIZE));
    output = [output; o.output_buffer];
    
    pi_cnt = [pi_cnt o.pi_cnt];
    sin_term = [sin_term o.sin_term];
    wr_ptr = [wr_ptr o.wr_ptr-1];
    delay_buf = [delay_buf; o.delay_buffer];
    in_out = [in_out; in(i*BLOCKSIZE - BLOCKSIZE + 1:i*BLOCKSIZE)];
    
    pi_reset = [pi_reset o.pi_reset];
    sin_inc = [sin_inc o.sin_inc];
    time_shift_idx = [time_shift_idx o.time_shift_idx];
    n = [n o.n];
end
% output = output./max(output);
% figure; plot(output);
% audiowrite('flanger.wav', output,fs)
fwrite(fo0, output, 'float32');

fwrite(fo_pi_cnt, pi_cnt, 'int');
fwrite(fo_sin_term, sin_term, 'float32');
fwrite(fo_wr_ptr, wr_ptr, 'int');
fwrite(fo_delay_buf, delay_buf, 'float32');
fwrite(fo_in_out, in_out, 'float32');
fwrite(fo_pi_reset, pi_reset, 'int');
fwrite(fo_sin_inc, sin_inc, 'float32');
fwrite(fo_time_shift_idx, time_shift_idx, 'int');
fwrite(fo_n, n, 'int');
fclose('all');

